cmd/virt-chroot/cgroup.go

```diff
@@ -0,0 +1,121 @@
+	}
+
+	err = json.Unmarshal(marshalledResources, &unmarshalledResources)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshall cgroups resources. "+
+			"marshalled resources: %s. err: %v", marshalledResources, err)
+	}
+
+	return &unmarshalledResources, err
+}
+
+func decodePaths(marshalledPathsHash string) (map[string]string, error) {
+	var unmarshalledPaths map[string]string
+
+	marshalledPaths, err := base64.StdEncoding.DecodeString(marshalledPathsHash)
+	if err != nil {
+		return nil, fmt.Errorf("cannot decode marshalled cgroups paths. "+
+			"encoded paths: %s. err: %v", marshalledPathsHash, err)
+	}
+
+	err = json.Unmarshal(marshalledPaths, &unmarshalledPaths)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshall cgroups paths. "+
+			"marshalled paths: %s. err: %v", marshalledPaths, err)
+	}
+
+	return unmarshalledPaths, err
+}
+
+func setCgroupResources(paths map[string]string, resources *runc_configs.Resources, isRootless bool, isV2 bool) error {
+	config := &runc_configs.Cgroup{
+		Path:      cgroup.HostCgroupBasePath,
+		Resources: &runc_configs.Resources{},
+	}
+
+	var err error
+
+	if isV2 {
+		err = setCgroupResourcesV2(paths, resources, isRootless, config)
+	} else {
+		err = setCgroupResourcesV1(paths, resources, isRootless, config)
+	}
+
+	if err != nil {
+		return fmt.Errorf("cannot set cgroup resources. err: %v", err)
+	}
+
+	return nil
+}
+
+func setCgroupResourcesV1(paths map[string]string, resources *runc_configs.Resources, isRootless bool, config *runc_configs.Cgroup) error {
+	return RunWithChroot(cgroup.HostCgroupBasePath, func() error {
+		cgroupManager := runc_fs.NewManager(config, paths, isRootless)
+		return cgroupManager.Set(resources)
+	})
+}
+
+func setCgroupResourcesV2(paths map[string]string, resources *runc_configs.Resources, isRootless bool, config *runc_configs.Cgroup) error {
+	cgroupDirPath := paths[""]
+
+	cgroupManager, err := runc_fs2.NewManager(config, cgroupDirPath, isRootless)
+	if err != nil {
+		return fmt.Errorf("cannot create cgroups v2 manager. err: %v", err)
+	}
+
+	err = cgroupManager.Set(resources)
+	return err
+}
+
+// RunWithChroot changes the root directory (via "chroot") into newPath, then
+// runs toRun function. When the function finishes, changes back the root directory
+// to the original one that
+func RunWithChroot(newPath string, toRun func() error) error {
+	// Ensure no other goroutines are effected by this
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	originalRoot, err := os.Open("/")
+	if err != nil {
+		return fmt.Errorf("failed to run with chroot - failed to open root directory. error: %v", err)
+	}
+
+	err = syscall.Chroot(newPath)
+	if err != nil {
+		return fmt.Errorf("failed to chroot into \"%s\". error: %v", newPath, err)
+	}
+
+	changeRootToOriginal := func() {
+		_ = originalRoot.Chdir()
+		_ = syscall.Chroot(".")
+	}
+	defer changeRootToOriginal()
+	defer originalRoot.Close()
```

Comment:
The defered calls are executed in last-in-first-out order. Here it is important that `defer originalRoot.Close()` is put before `defer changeRootToOriginal()`. Otherwise `_ = originalRoot.Chdir()` will return an error (which is ignored currerently).

Reply:
Thanks for catching that! Important info.. I will remember it from now on.

To avoid any errors in the future I moved `originalRoot.Close()` into `changeRootToOriginal()` function. This way order of execution is obvious and less error-prone